subsys/bluetooth/services/gadgets.c

```diff
@@ -0,0 +1,525 @@
+		if (err) {
+			LOG_WRN("Failed to send continuation packet");
+			transmit_success = false;
+		} else {
+			return;
+		}
+	}
+
+	atomic_set(&transmit_busy, false);
+
+	if (gadgets_cb.sent_cb) {
+		gadgets_cb.sent_cb(
+			conn,
+			/* In case of fragmented transmission */
+			user_data == tx_buffer ? stream_payload : user_data,
+			transmit_success);
+	}
+}
+
+static ssize_t on_cccd_write(
+	struct bt_conn *conn, const struct bt_gatt_attr *attr, uint16_t value)
+{
+	if (gadgets_cb.cccd_cb) {
+		bool notif_enabled;
+
+		notif_enabled = (value & BT_GATT_CCC_NOTIFY) != 0;
+
+		gadgets_cb.cccd_cb(notif_enabled);
+	}
+
+	return sizeof(value);
+}
+
+/* GADGETS Service Declaration */
+/* RX data direction = from this Service to GATT Client */
+/* TX data direction  = from GATT Client to this Service */
+BT_GATT_SERVICE_DEFINE(gadgets_svc,
+BT_GATT_PRIMARY_SERVICE(BT_UUID_GADGETS_SERVICE),
+	BT_GATT_CHARACTERISTIC(
+		BT_UUID_GADGETS_RX,
+		BT_GATT_CHRC_NOTIFY | BT_GATT_CHRC_READ,
+		BT_GATT_PERM_READ_ENCRYPT,
+		NULL, NULL, NULL),
+	BT_GATT_CCC_MANAGED(
+		((struct _bt_gatt_ccc[])
+			{BT_GATT_CCC_INITIALIZER(NULL, on_cccd_write, NULL)}),
+			(BT_GATT_PERM_READ_ENCRYPT | BT_GATT_PERM_WRITE_ENCRYPT)),
+	BT_GATT_CHARACTERISTIC(
+		BT_UUID_GADGETS_TX,
+		BT_GATT_CHRC_WRITE |
+		BT_GATT_CHRC_WRITE_WITHOUT_RESP,
+		BT_GATT_PERM_READ_ENCRYPT | BT_GATT_PERM_WRITE_ENCRYPT,
+		NULL, on_receive, NULL),
+);
+
+static int bt_gatt_gadgets_send_perform(
+			struct bt_conn *conn, void *data, size_t size)
+{
+	const struct bt_gatt_attr *attr = &gadgets_svc.attrs[2];
+	struct bt_gatt_notify_params params = {
+		.attr = attr,
+		.data = data,
+		.len = size,
+		.func = on_sent,
+		.user_data = data,
+	};
+
+	return bt_gatt_notify_cb(conn, &params);
+}
+
+int bt_gatt_gadgets_send_stream(
+		struct bt_conn *conn,
+		enum stream_id stream,
+		void *data,
+		size_t size)
+{
+	uint16_t max_packet_size;
+	size_t payload_size;
+	size_t header_size;
+	int err;
+
+	if (conn == NULL) {
+		return -EINVAL;
+	}
+
+	if (atomic_set(&transmit_busy, true)) {
+		return -EBUSY;
+	}
+
+	tx_sequence_number = 0;
+
+	max_packet_size = bt_gatt_gadgets_max_send(conn);
+	stream_payload = data;
+
+	header_size = sizeof(struct gadgets_header_first);
+	if ((header_size + size) > max_packet_size) {
+		stream_payload_remainder = size + header_size - max_packet_size;
+	} else {
+		stream_payload_remainder = 0;
+	}
+
+	payload_size = size - stream_payload_remainder;
+	stream_payload_offset = payload_size;
+
+	encode_header(
+		&tx_buffer[0],
+		stream,
+		GADGETS_TRANSACTION_TYPE_FIRST,
+		payload_size,
+		size);
+	memcpy(&tx_buffer[header_size], data, payload_size);
+
+	err = bt_gatt_gadgets_send_perform(
+			conn, tx_buffer, header_size + payload_size);
+	if (err) {
+		atomic_set(&transmit_busy, false);
+	}
+
+	return err;
+}
+
+int bt_gatt_gadgets_send(struct bt_conn *conn, void *data, size_t size)
+{
+	int err;
+
+	if (atomic_set(&transmit_busy, true)) {
+		return -EBUSY;
+	}
+
```

Comment:
shouldn't you set ```stream_payload_remainder``` to 0 here? In case, there are some state leftovers from stream-based send method?

``on_sent`` callback is triggered also for this simple send method

Reply:
If transmit_busy is false and remainder is non-zero, there is an error somewhere. I can add an ASSERT here to detect such error intead?

Reply:
yes, sure. ASSERT seems fine